presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/KdbTree.java

```diff
@@ -0,0 +1,358 @@
+
+    private static Node buildKdbTreeNode(int maxItemsPerNode, int level, Rectangle extent, List<Rectangle> items, LeafIdAllocator leafIdAllocator)
+    {
+        checkArgument(maxItemsPerNode > 0, "maxItemsPerNode must be > 0");
+        checkArgument(level >= 0, "level must be >= 0");
+        checkArgument(level <= MAX_LEVELS, "level must be <= 10,000");
+        requireNonNull(extent, "extent is null");
+        requireNonNull(items, "items is null");
+
+        if (items.size() <= 1 || items.size() <= maxItemsPerNode || level == MAX_LEVELS) {
+            return newLeaf(extent, leafIdAllocator.next());
+        }
+
+        // Split over longer side
+        boolean splitVertically = extent.getWidth() >= extent.getHeight();
+        Optional<SplitResult<Node>> splitResult = trySplit(splitVertically ? BY_X : BY_Y, maxItemsPerNode, level, extent, items, leafIdAllocator);
+        if (!splitResult.isPresent()) {
+            // Try spitting by the other side
+            splitResult = trySplit(splitVertically ? BY_Y : BY_X, maxItemsPerNode, level, extent, items, leafIdAllocator);
+        }
+
+        if (!splitResult.isPresent()) {
+            return newLeaf(extent, leafIdAllocator.next());
+        }
+
+        return newInternal(extent, splitResult.get().getLeft(), splitResult.get().getRight());
+    }
+
+    private static final class SplitResult<T>
+    {
+        private final T left;
+        private final T right;
+
+        private SplitResult(T left, T right)
+        {
+            this.left = requireNonNull(left, "left is null");
+            this.right = requireNonNull(right, "right is null");
+        }
+
+        public T getLeft()
+        {
+            return left;
+        }
+
+        public T getRight()
+        {
+            return right;
+        }
+    }
+
+    private static Optional<SplitResult<Node>> trySplit(SplitDimension splitDimension, int maxItemsPerNode, int level, Rectangle extent, List<Rectangle> items, LeafIdAllocator leafIdAllocator)
+    {
+        checkArgument(items.size() > 1, "Number of items to split must be > 1");
+
+        // Sort envelopes by xMin or yMin
+        List<Rectangle> sortedItems = ImmutableList.sortedCopyOf(splitDimension.getComparator(), items);
+
+        // Find a mid-point
+        int middleIndex = (sortedItems.size() - 1) / 2;
+        Rectangle middleEnvelope = sortedItems.get(middleIndex);
+        double splitValue = splitDimension.getValue(middleEnvelope);
+        int splitIndex = middleIndex;
+
+        // skip over duplicate values
+        while (splitIndex < sortedItems.size() && splitDimension.getValue(sortedItems.get(splitIndex)) == splitValue) {
+            splitIndex++;
+        }
+
+        // all values between left-of-middle and the end are the same, so can't split
+        if (splitIndex == sortedItems.size()) {
+            return Optional.empty();
+        }
+
+        splitValue = (splitValue + splitDimension.getValue(sortedItems.get(splitIndex))) / 2;
```

Comment:
This needs a comment, as it's not entirely obvious why it's needed (per my previous question).

Reply:
@martint How about this?

```
        // about half of the objects are <= splitValue, the rest are >= next value
        // assuming the input set of objects is a sample from a much larger set, 
        // let's split in the middle; this way objects from the larger set with values 
        // between splitValue and next value will get split somewhat evenly into left 
        // and right partitions
        splitValue = (splitValue + splitDimension.getValue(sortedItems.get(splitIndex))) / 2;
```